### PR TITLE
Adopt the new open-vm-tools test in MU incident testing

### PR DESCRIPTION
**Pull request by nanzhg**
1. Support qemu backend for MU incident tests.
2. Update ssh_hypervisor_init.pm to backup ssh keys for connecting vmware server.

- Related ticket: https://progress.opensuse.org/issues/103266
- Verification run: 
   sle12sp3 - https://openqa.suse.de/tests/10698228
   sle12sp4 - https://openqa.suse.de/tests/10698232
   sle12sp5 - https://openqa.suse.de/tests/10701678
   sle15sp1 - https://openqa.suse.de/tests/10698229
   sle15sp2 - https://openqa.suse.de/tests/10698219
   sle15sp3 - https://openqa.suse.de/tests/10698218
   sle15sp4 - https://openqa.suse.de/tests/10702262
   post_fail_hook - https://openqa.suse.de/tests/10715902#step/esxi_open_vm_tools/71

### CODE REVIEW

**Comment by nanzhg** [comment]
@alice-suse @xguo @waynechen55 @Julie-CAO @tbaev @RoyCai7 @vojha please help to review this PR. Thanks

**Comment by RoyCai7** [comment]
It is better to run a job that use svirt backend.

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 1]
```diff
@@ -11,14 +11,28 @@ use strict;
...
+my $VM_POWER_OFF = 'Powered off';
 
 sub run {
-    my $vm_name = console('svirt')->name;
+    if (check_var('BACKEND', 'svirt')) {
+        my $vm_name = console('svirt')->name;
+        run_tests($vm_name);
+    }
+    elsif (check_var('BACKEND', 'qemu')) {
+        foreach my $vm_name (keys %virt_autotest::common::guests) {
```
Comment (alice-suse) [inline tests/virt_autotest/esxi_open_vm_tools.pm:28]: I am afraid this is NOT the logic for MU. See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/virtualization/universal/patch_guests.pm#L43,  which is applied in job https://openqa.suse.de/tests/10497782/modules/patch_guests/steps/1/src. Only the guest with the same version of product with incident, will need to be tested.

File: lib/virt_autotest/esxi_utils.pm [hunk 2]
```diff
@@ -73,13 +114,38 @@ sub disable_vm_time_synchronization {
...
+        console('svirt')->run_cmd("echo time.synchronize.resume.host=\"FALSE\" >> $vmx_file", domain => 'sshVMwareServer', wantarray => 1);
+    }
+    elsif (check_var('BACKEND', 'qemu')) {
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "sed -ie 's/tools.syncTime.*/tools.syncTime=\"FALSE\"/' $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.continue=\"FALSE\" >> $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.restore=\"FALSE\" >> $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.resume.disk=\"FALSE\" >> $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.shrink=\"FALSE\" >> $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.tools.startup=\"FALSE\" >> $vmx_file"));
+        script_run(qq(ssh -o StrictHostKeyChecking=no root\@$hypervisor "echo time.synchronize.resume.host=\"FALSE\" >> $vmx_file"));
```
Comment (alice-suse) [inline lib/virt_autotest/esxi_utils.pm:133]: Will it be problematic if some of above script_run commands fail? If yes, maybe switching to assert_script_run?

File: lib/virt_autotest/esxi_utils.pm [hunk 3]
```diff
@@ -73,13 +114,38 @@ sub disable_vm_time_synchronization {
...
+    }
+}
+
+sub recover_vmx_config {
```
Comment (alice-suse) [inline lib/virt_autotest/esxi_utils.pm:137]: Seems this function only reset time related config, not all config? How about making the function name more specific to reflect this?

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 4]
```diff
@@ -137,16 +164,34 @@ sub get_epoch_time {
 sub init_guest_time {
     my $h_datetime = get_host_timestamp();
     # Get last day of the host time
-    my $last_day = script_output("date -u -d '$h_datetime last day' +'\%F \%T'");
+    my $last_day;
+    if (check_var('BACKEND', 'svirt')) {
+        $last_day = script_output("date -u -d '$h_datetime last day' +'\%F \%T'");
+    }
+    elsif (check_var('BACKEND', 'qemu')) {
+        $last_day = script_output(qq($ssh_vm "date -u -d '$h_datetime last day' +'\%F \%T'"));
```
Comment (alice-suse) [inline tests/virt_autotest/esxi_open_vm_tools.pm:172]: It seems this cmd can work for both backends, why differentiate?

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 5]
```diff
@@ -137,16 +164,34 @@ sub get_epoch_time {
...
+    }
     # Set the guest time by using the variable $last_day
-    my $g_datetime = script_output("date -u -s '$last_day' +'\%F \%T'");
+    my $g_datetime;
+    if (check_var('BACKEND', 'svirt')) {
+        $g_datetime = script_output("date -u -d '$h_datetime last day' +'\%F \%T'");
+    }
+    elsif (check_var('BACKEND', 'qemu')) {
+        $g_datetime = script_output(qq($ssh_vm "date -u -s '$last_day' +'\%F \%T'"));
```
Comment (alice-suse) [inline tests/virt_autotest/esxi_open_vm_tools.pm:180]: Same here? Same cmd is ok?

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 6]
```diff
@@ -137,16 +164,34 @@ sub get_epoch_time {
...
 
 sub get_diff_seconds {
     my $h_datetime = get_host_timestamp();
-    my $g_datetime = script_output("date -u +'\%F \%T'");
+    my $g_datetime;
+    if (check_var('BACKEND', 'svirt')) {
+        $g_datetime = script_output("date -u +'\%F \%T'");
+    }
+    elsif (check_var('BACKEND', 'qemu')) {
+        $g_datetime = script_output(qq($ssh_vm "date -u +'\%F \%T'"));
```
Comment (alice-suse) [inline tests/virt_autotest/esxi_open_vm_tools.pm:193]: So does here?

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 7]
```diff
@@ -268,6 +313,19 @@ sub do_clock_sync_tests {
...
         die 'Disabling clock synchronization failed.';
     }
+
+    if (check_var('BACKEND', 'qemu')) {
+        record_info('Recover vmx file configuration');
+        $powerops = 'power.shutdown';
+        $powerops_ret = take_vm_power_ops($vm_id, $powerops, $VM_POWER_ON);
+        check_vm_power_state($vm_id, $vm_ip, $powerops, $powerops_ret, 0, $VM_POWER_OFF);
+
+        recover_vmx_config($vm_name);
```
Comment (alice-suse) [inline tests/virt_autotest/esxi_open_vm_tools.pm:323]: Careful consideration here :+1:

**Comment by alice-suse** [comment]
> It is better to run a job that uses svirt backend.

Agree, and I would suggest to add below verification tests:
- qemu backend : 
  - PASS job with patch_guests.pm , just as normal MU jobs do in https://openqa.suse.de/tests/10497782
  -  Fail job (can fail on purpose on some line to check post_fail_hook)
- svirt backend: both pass and fail to do regression check

**Comment by nanzhg** [comment]
> It is better to run a job that uses svirt backend.

Agree. But svirt part run with textmode_svirt, it's blocked by defect poo#124161.

File: lib/virt_autotest/esxi_utils.pm [hunk 8]
```diff
@@ -25,46 +25,84 @@ our @EXPORT = qw(
...
 
+my $hypervisor = get_var('HYPERVISOR') // 'esxi7.qa.suse.cz';
+
 sub esxi_vm_get_vmid {
     my $vm_name = shift;
-    my $vim_cmd = "vim-cmd vmsvc/getallvms | grep -i $vm_name | awk -F' ' '{print \$1}'";
-    my (undef, $vmid) = console('svirt')->run_cmd($vim_cmd, domain => 'sshVMwareServer', wantarray => 1);
-    chomp($vmid);
+    my $vim_cmd = "vim-cmd vmsvc/getallvms | grep -i $vm_name | cut -d' ' -f1";
+    my $vmid;
```
Comment (waynechen55) [inline lib/virt_autotest/esxi_utils.pm:36]: Initialize ```my $vmid = 0 or ''``` ?

File: tests/virt_autotest/esxi_open_vm_tools.pm [hunk 9]
```diff
@@ -101,9 +122,15 @@ sub wait_for_vm_network {
...
+    do {
+        if (is_svirt) {
+            $is_nw_up = console('svirt')->run_cmd($cmd);
+        }
+        elsif (is_qemu) {
+            $is_nw_up = script_run($cmd);
+        }
         $times--;
-    }
+    } while ($times && $is_nw_up);
```
Comment (waynechen55) [inline tests/virt_autotest/esxi_open_vm_tools.pm:133]: Any follow-up actions after this "monitor" block ?

File: lib/virt_autotest/esxi_utils.pm (hunk 2 above)
Comment (nanzhg) [reply lib/virt_autotest/esxi_utils.pm:130]: Yes, should use assert_script_run. Thanks for pointing out!

File: lib/virt_autotest/esxi_utils.pm (hunk 3 above)
Comment (nanzhg) [reply lib/virt_autotest/esxi_utils.pm:134]: How about the name 'revert_vm_timesync_setting'?

File: tests/virt_autotest/esxi_open_vm_tools.pm (hunk 4 above)
Comment (nanzhg) [reply tests/virt_autotest/esxi_open_vm_tools.pm:172]: Yes, it doesn't work by some special characters with ssh which post a failure.

File: tests/virt_autotest/esxi_open_vm_tools.pm (hunk 5 above)
Comment (nanzhg) [reply tests/virt_autotest/esxi_open_vm_tools.pm:180]: Refer to https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/16179#discussion_r1116829329

File: tests/virt_autotest/esxi_open_vm_tools.pm (hunk 6 above)
Comment (nanzhg) [reply tests/virt_autotest/esxi_open_vm_tools.pm:193]: Refer to https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/16179#discussion_r1116829329

File: lib/virt_autotest/esxi_utils.pm (hunk 8 above)
Comment (nanzhg) [reply lib/virt_autotest/esxi_utils.pm:36]: It declares the variable $vmid, and the default value is 'undef', isn't it?

File: tests/virt_autotest/esxi_open_vm_tools.pm (hunk 9 above)
Comment (nanzhg) [reply tests/virt_autotest/esxi_open_vm_tools.pm:134]: Not quite understand. This block will be repeatedly called as per the test scenario. The only change in this function is that added a condition to judge whether qemu or svirt backend.

**Comment by nanzhg** [comment]
> > It is better to run a job that uses svirt backend.
> 
> Agree, and I would suggest to add below verification tests:
> 
> * qemu backend :
>   
>   * PASS job with patch_guests.pm , just as normal MU jobs do in https://openqa.suse.de/tests/10497782
>   * Fail job (can fail on purpose on some line to check post_fail_hook)
> * svirt backend: both pass and fail to do regression check

Sure, once a new build is available, I'll trigger a verification run with patch_guests.pm and also check the failure behavior. For svirt backend, it requires the defect poo#124161 fixed.

File: schedule/virtualization/vmware_hyperv.yaml [hunk 10]
```diff
@@ -10,9 +10,8 @@ schedule:
     - virtualization/universal/upgrade_guests
     - virtualization/universal/patch_guests
     - '{{open_vm_tools}}'
-
 conditional_schedule:
     open_vm_tools:
         REGRESSION:
             vmware:
-                - virtualization/universal/open_vm_tools
+                - virt_autotest/esxi_open_vm_tools
```
Comment (alice-suse) [inline schedule/virtualization/vmware_hyperv.yaml:17]: @nanzhg  Hi Nan, based on final agreement with PJMs in description of https://progress.opensuse.org/issues/103266, for sle12sp5 and sle15sp4+, new test is used, while for older products sle15sp[1-3] and sle12sp[2-4] , the old test is used. So will need a `if else` here.

Comment (nanzhg) [reply schedule/virtualization/vmware_hyperv.yaml:17]: Hi Alice, I've made the changes, please review. Thanks

Comment (alice-suse) [reply schedule/virtualization/vmware_hyperv.yaml:37]: Thanks @nanzhg . It looks good this way. BTW, sle15 is not included in the final confirmed product matrix. Please refer to the ticket's description for final requirement.

File: lib/virt_autotest/esxi_utils.pm (hunk 8 above)
Comment (alice-suse) [reply lib/virt_autotest/esxi_utils.pm:37]: @nanzhg  Hi Nan, you can have a look at https://progress.opensuse.org/issues/120750. We used to have dedicated ticket to fix uninitialized values to stop perl warnings. Recommend to follow this good habit.

Comment (nanzhg) [reply lib/virt_autotest/esxi_utils.pm:37]: @alice-suse OK, there is only one warning I can see in the following job log, and I've fixed it.

https://openqa.suse.de/tests/10621486/logfile?filename=autoinst-log.txt
`Use of uninitialized value in numeric eq (==) at /var/lib/openqa/pool/3/os-autoinst-distri-opensuse/tests/virt_autotest/esxi_open_vm_tools.pm line 106.`

Warning statement:
`    } elsif ($powerops_ret == undef) {`
Fixed with:
`    } elsif (!defined $powerops_ret) {`

**Comment by alice-suse** [comment]
Hi Nan, great work!
BTW, as commented earlier, it is better to add failure jobs (triggered on purpose) to see if post_fail_hood work as expected. Besides, in  https://progress.opensuse.org/issues/103266#note-44, this needs the validation for open-vm-tools incidents.  I do not see job links given for that. Do I miss them anywhere?

I suppose that there won't be open-vm-tools incident soon. So it is ok to merge this PR, but keep the progress ticket open. Once that's doable, the left work can be done and if needed further minor fix PR for it can be opened newly, then poo is resolved after that.  HDYT?

**Comment by nanzhg** [comment]
> Hi Nan, great work! BTW, as commented earlier, it is better to add failure jobs (triggered on purpose) to see if post_fail_hood work as expected. Besides, in https://progress.opensuse.org/issues/103266#note-44, this needs the validation for open-vm-tools incidents. I do not see job links given for that. Do I miss them anywhere?
> 
> I suppose that there won't be open-vm-tools incident soon. So it is ok to merge this PR, but keep the progress ticket open. Once that's doable, the left work can be done and if needed further minor fix PR for it can be opened newly, then poo is resolved after that. HDYT?

Yes, all validation runs are based on kernel-default incidents, and the link for post failure check is updated. Please review.

**Comment by alice-suse** [comment]
Will merge it after CI check passes, if no more comments from anyone else today.

**Comment by guoxuguang** [comment]
LGTM

**Pull request merged**